File: Beam/Source/QueriesTests/SequenceTester.cpp
#include <sstream>
#include <doctest/doctest.h>
#include "Beam/Queries/Sequence.hpp"
#include "Beam/SerializationTests/ValueShuttleTests.hpp"

using namespace Beam;
using namespace Beam::Tests;
using namespace boost;
using namespace boost::gregorian;
using namespace boost::posix_time;

TEST_SUITE("Sequence") {
  TEST_CASE("default_constructor") {
    auto sequence = Sequence();
    REQUIRE(sequence.get_ordinal () == 0);
  }

  TEST_CASE("ordinal_constructor") {
    auto sequence_a = Sequence(1);
    REQUIRE(sequence_a.get_ordinal () == 1);
    auto sequence_b = Sequence(5);
    REQUIRE(sequence_b.get_ordinal () == 5);
    auto sequence_c = Sequence(std::numeric_limits<std::uint64_t>::max());
    REQUIRE(
      sequence_c.get_ordinal () == std::numeric_limits<std::uint64_t>::max());
  }

  TEST_CASE("first_sequence") {
    auto first = Sequence::FIRST;
    REQUIRE(first.get_ordinal () == 0);
  }

  TEST_CASE("last_sequence") {
    auto last = Sequence::LAST;
    REQUIRE(last.get_ordinal () == std::numeric_limits<std::uint64_t>::max());
  }

  TEST_CASE("less_than_operator") {
    REQUIRE(!(Sequence(0) < Sequence(0)));
    REQUIRE(Sequence(0) < Sequence(1));
    REQUIRE(!(Sequence(1) < Sequence(0)));
    REQUIRE(!(Sequence(1) < Sequence(1)));
  }

  TEST_CASE("less_than_or_equal_operator") {
    REQUIRE(Sequence(0) <= Sequence(0));
    REQUIRE(Sequence(0) <= Sequence(1));
    REQUIRE(!(Sequence(1) <= Sequence(0)));
    REQUIRE(Sequence(1) <= Sequence(1));
  }

  TEST_CASE("equals_operator") {
    REQUIRE(Sequence(0) == Sequence(0));
    REQUIRE(!(Sequence(0) == Sequence(1)));
    REQUIRE(!(Sequence(1) == Sequence(0)));
    REQUIRE(Sequence(1) == Sequence(1));
  }

  TEST_CASE("not_equals_operator") {
    REQUIRE(!(Sequence(0) != Sequence(0)));
    REQUIRE(Sequence(0) != Sequence(1));
    REQUIRE(Sequence(1) != Sequence(0));
    REQUIRE(!(Sequence(1) != Sequence(1)));
  }

  TEST_CASE("greater_than_or_equal_operator") {
    REQUIRE(Sequence(0) >= Sequence(0));
    REQUIRE(!(Sequence(0) >= Sequence(1)));
    REQUIRE(Sequence(1) >= Sequence(0));
    REQUIRE(Sequence(1) >= Sequence(1));
  }

  TEST_CASE("greater_than_operator") {
    REQUIRE(!(Sequence(0) > Sequence(0)));
    REQUIRE(!(Sequence(0) > Sequence(1)));
    REQUIRE(Sequence(1) > Sequence(0));
    REQUIRE(!(Sequence(1) > Sequence(1)));
  }

  TEST_CASE("increment") {
    REQUIRE(increment(Sequence(0)) == Sequence(1));
    REQUIRE(increment(Sequence(1)) == Sequence(2));
    REQUIRE(increment(Sequence(2)) == Sequence(3));
    REQUIRE(increment(Sequence::LAST) == Sequence::LAST);
  }

  TEST_CASE("decrement") {
    REQUIRE(decrement(Sequence(1)) == Sequence(0));
    REQUIRE(decrement(Sequence(2)) == Sequence(1));
    REQUIRE(decrement(Sequence(3)) == Sequence(2));
    REQUIRE(decrement(Sequence::FIRST) == Sequence::FIRST);
  }

  TEST_CASE("encoding_timestamp") {
    auto timestamp =
      ptime(date(1984, May, 6), hours(12) + minutes(44) + seconds(53));
    auto sequence = to_sequence(timestamp);
    auto encoding = static_cast<Sequence::Ordinal>(0b00111110000000101000110) <<
      (CHAR_BIT * sizeof(Sequence::Ordinal) - 23);
    REQUIRE(encoding == sequence.get_ordinal ());
    REQUIRE(decode_timestamp(Sequence(encoding)).date() == timestamp.date());
  }

  TEST_CASE("stream") {
    auto ss = std::stringstream();
    ss << Sequence(123);
    REQUIRE(ss.str() == "123");
    test_round_trip_shuttle(Sequence(543));
  }
}

File: WebApi/tests/queries/sequence_tester.ts
import { describe, it } from 'node:test';
import * as assert from 'node:assert';
import { Sequence } from '../../source/queries/sequence';

describe('Sequence', () => {
  it('default_constructor', () => {
    const sequence = new Sequence();
    assert.strictEqual(sequence.ordinal, 0);
    assert.ok(sequence.equals(Sequence.FIRST));
  });

  it('increment', () => {
    assert.strictEqual(new Sequence(5).increment().ordinal, 6);
    assert.ok(Sequence.LAST.increment().equals(Sequence.LAST));
  });

  it('decrement', () => {
    assert.strictEqual(new Sequence(5).decrement().ordinal, 4);
    assert.ok(Sequence.FIRST.decrement().equals(Sequence.FIRST));
  });

  it('compare', () => {
    assert.ok(new Sequence(1).compare(new Sequence(2)) < 0);
    assert.ok(new Sequence(2).compare(new Sequence(1)) > 0);
    assert.strictEqual(new Sequence(1).compare(new Sequence(1)), 0);
  });

  it('equals', () => {
    assert.ok(new Sequence(10).equals(new Sequence(10)));
    assert.ok(!new Sequence(10).equals(new Sequence(11)));
  });

  it('to_string', () => {
    assert.strictEqual(new Sequence(123).toString(), '123');
  });

  it('round_trip', () => {
    const sequence = new Sequence(42);
    assert.strictEqual(sequence.toJson(), 42);
    assert.ok(Sequence.fromJson(sequence.toJson()).equals(sequence));
  });
});
